engine/src/main/java/io/camunda/zeebe/engine/state/migration/DbMigrationController.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.engine.state.migration;
+
+import io.camunda.zeebe.engine.processing.streamprocessor.ReadonlyProcessingContext;
+import io.camunda.zeebe.engine.processing.streamprocessor.StreamProcessorLifecycleAware;
+import io.camunda.zeebe.engine.state.mutable.MutableZeebeState;
+import java.util.function.Function;
+
+public final class DbMigrationController implements StreamProcessorLifecycleAware {
+
+  private DbMigrator dbMigrator;
+  private final Function<MutableZeebeState, DbMigrator> migratorFactory;
+
+  public DbMigrationController() {
+    this(DbMigratorImpl::new);
+  }
+
+  protected DbMigrationController(final Function<MutableZeebeState, DbMigrator> migratorFactory) {
+    this.migratorFactory = migratorFactory;
+  }
+
+  @Override
+  public final void onRecovered(final ReadonlyProcessingContext context) {
+    final var migrator = migratorFactory.apply(context.getZeebeState());
+
+    synchronized (this) {
+      dbMigrator = migrator;
+    }
+    try {
+      migrator.runMigrations();
```

Comment:
I need a bit more context here possibly, so I apologize in advance if I mention some things which are wrong :sweat_smile: 

Would this block only the stream processor actor? In this case, it may cause it to become unhealthy and fail. This would recover, but it would still look strange - afaik we don't really react to it become unhealthy, just report it. So it's more of an observability issue. Other problems may be if you're already receiving traffic, it delays the time until you can process commands, but I think that's not really avoidable right now, and out of scope here. It could also cause a build up of timers, I suppose, but I don't think that's all that likely for more use cases. So I think the biggest downside is that it will mark the partition as unhealthy for however long it takes to run, which isn't really true. One possibility is introducing a new processing phase where it's migrating, in which case the health of the component can be reported differently. You might still want to break it up into multiple steps however if only to detect livelock, but again that might be out of scope.

Reply:
As far as I know this would block the stream processor actor.

I agree with Philipp that this can be dangerous in the general case. However, for the migrations for 1.1.0 I don't think it's a problem yet.

I strongly believe it needs to be a synchronous process and strictly before any processing happens.

Long term, I would prefer to see the migration logic being triggered as part of the rolling update process directly instead of during recovery.

Reply:
> As far as I know this would block the stream processor actor.

Yes. To be more concrete, it would block the whole actor thread. Within the migration, it does everything synchronously without calling `actor.run()` or `actor.submit()`. In our actor model, this is not recommended. 

If the broker has only one CPU then the migration may block the whole system. Right? 

In the long run, we should introduce a separate phase before processing, or externalize it.

In this case, it should be okay to do it this way.

Reply:
That's true - generally we should avoid long running, blocking operation, and split them into chunks of work. How easy would it be here? Would we have to make the migrations themselves actor-aware? I think that's probably the best approach in the long term to be honest - that is, that the migrations are not a single blocking operations but many smaller chunks which can be executed, interleaved with other actors.

I don't necessarily want to block progress right now, as I think that we can still extend what's here to allow this, no? Correct me if I'm wrong, in which case we should do it now.

Reply:
I'm okay with creating a follow-up issue. This issue should be done before adding the next migration step :sweat_smile:

Reply:
I am confused. 

We have two migrations ready to be reviewed. The two migrations we want to add for 1.1.0 will be in the ms range. 

I think your concerns are valid in the general case, but not so applicable to the problem at hand.

Should the new issue be done before that or after the migrations we have already implemented?

Also I cannot imagine doing the migration in an actor aware way - unless there is a way to make the rest of the the stream processing wait until migrations are completely run.

I think to do it right, we need some kind of rolling update process which triggers migrations in a defined way, without boot strapping anything else of the broker.

This can be done, of course, but it is a long way from the start of all this, which was https://github.com/camunda-cloud/zeebe/issues/6364, which was a "nice to have" topic pre GA. 

So please be a little more explicit on what your expectation are on how to proceed, and in what sequence.

Reply:
From my side this is fine to merge as is, and we don't need to work on this immediately, but I would like a follow up issue to address this in the future. In the mean time, what is the entry point where someone would add/define a new migration? In `DbMigratorImpl`? We can add a comment there that reminds us to be careful if we have long running migrations, link to the follow up issue and mention that it might be necessary to implement that if we do have a long running, blocking migration task.

Regarding your question about the stream processing waiting - we can pause the stream processor, we already have this feature, so we could eventually reuse this. Pause and unpause it only once the migrations have finished.

Reply:
Added a new issue: https://github.com/camunda-cloud/zeebe/issues/7248
Feel free to add any comments, or thoughts there. I also reference this issue in the code base at suitable locations.